org.eclipse.lemminx/src/main/java/com/thaiopensource/relaxng/pattern/CMRelaxNGElementDeclaration.java

```diff
@@ -233,15 +234,19 @@ public boolean isMixedContent() {
 	@Override
 	public boolean isOptional(String childElementName) {
 		if (requiredElementNames == null) {
-			RequiredElementsFunction elementsFunction = new RequiredElementsFunction();
-			requiredElementNames = pattern.apply(elementsFunction);
-			for (Name name : requiredElementNames) {
-				if (childElementName.equals(name.getLocalName())) {
-					return false;
+			requiredElementNames = new HashSet<>();

```

Comment:
After some investigation, we cannot use the standard RequiredElementsFunction  because it returns only the first required element. I suggest that you create thefollowing class:

```java
package com.thaiopensource.relaxng.pattern;

import java.util.Set;

public class MyRequiredElementsFunction extends AbstractPatternFunction<Void> {

	private final Set<String> requiredElementNames;

	public MyRequiredElementsFunction(Set<String> requiredElementNames) {
		this.requiredElementNames = requiredElementNames;
	}

	@Override
	public Void caseOther(Pattern p) {
		return null;
	}

	@Override
	public Void caseElement(ElementPattern p) {
		return caseNamed(p.getNameClass());
	}

	@Override
	public Void caseGroup(GroupPattern p) {
		return union(p);
	}

	@Override
	public Void caseChoice(ChoicePattern p) {
		// For choice, get the first pattern which is not nullable
		Pattern p1 = p.getOperand1();
		if (!p1.isNullable()) {
			p1.apply(this);
		} else {
			Pattern p2 = p.getOperand2();
			if (!p2.isNullable()) {
				p2.apply(this);
			}
		}
		return null;
	}

	@Override
	public Void caseInterleave(InterleavePattern p) {
		return union(p);
	}

	@Override
	public Void caseAfter(AfterPattern p) {
		return p.getOperand1().apply(this);
	}

	@Override
	public Void caseOneOrMore(OneOrMorePattern p) {
		return p.getOperand().apply(this);
	}

	private Void caseNamed(NameClass nc) {
		if (!(nc instanceof SimpleNameClass))
			return null;
		requiredElementNames.add(((SimpleNameClass) nc).getName().getLocalName());
		return null;
	}
	
	private Void union(BinaryPattern p) {
		Pattern p1 = p.getOperand1();
		if (!p1.isNullable()) {
			p1.apply(this);
		}
		Pattern p2 = p.getOperand2();
		if (!p2.isNullable()) {
			p2.apply(this);
		}
		return null;
	}
}

```

and replace isOptionalmethod like this:

```java
@Override
public boolean isOptional(String childElementName) {
	if (requiredElementNames == null) {
		requiredElementNames = new HashSet<>();
		MyRequiredElementsFunction elementsFunction = new MyRequiredElementsFunction(requiredElementNames);
			getPattern().getContent().apply(elementsFunction);
	}
	return !requiredElementNames.contains(childElementName);
}
```

For the usecase of TEI, it should generate only:

```xml
<teiHeader>
	<fileDesc>
		<titleStmt>
			<title>
			</title>
		</titleStmt>
		<publicationStmt>
			<publisher>
			</publisher>
		</publicationStmt>
		<sourceDesc>
			<p>
			</p>
		</sourceDesc>
	</fileDesc>
</teiHeader>
<text>
	<body>
		<div>
		</div>
	</body>
</text>
```